File: tools/i18n_missing.mjs
// Textos de «Tu cuenta» que no tienen traducción inglesa.
//
// Busca los t('…') de app.js y cuenta.js, y los textos de las listas que se
// traducen al vuelo (motivos, tipos de sugerencia, estados), y dice cuáles no
// están en APP_EN. Sale con 1 si falta alguno, para poder usarlo de prueba.
//
//   node tools/i18n_missing.mjs
import { readFileSync } from 'node:fs';
import vm from 'node:vm';

const raiz = new URL('../app/', import.meta.url);
const leer = (f) => readFileSync(new URL(f, raiz), 'utf8');

const ctx = {};
vm.runInNewContext(`${leer('i18n.js')}\nthis.APP_EN = APP_EN;`, ctx);
const EN = ctx.APP_EN;

const codigo = leer('app.js') + leer('cuenta.js');
const textos = new Set();
for (const m of codigo.matchAll(/\bt\(\s*'((?:[^'\\]|\\.)*)'\s*\)/g)) textos.add(m[1].replace(/\\'/g, "'"));
// Listas: ['clave', 'Texto', 'Ejemplo'] y { clave: 'Texto' } dentro de las
// constantes que se pasan por t().
for (const nombre of ['TIPOS_SUGERENCIA', 'MOTIVOS']) {
  const bloque = codigo.match(new RegExp(`const ${nombre} = \\[([\\s\\S]*?)\\n\\];`));
  if (!bloque) continue;
  for (const fila of bloque[1].matchAll(/\['[a-z_]+', '([^']*)'(?:, '([^']*)')?\]/g)) {
    if (fila[1]) textos.add(fila[1]);
    if (fila[2]) textos.add(fila[2]);
  }
}
for (const nombre of ['ESTADO_SUGERENCIA', 'QUE_SE_DENUNCIA', 'ERRORES']) {
  for (const bloque of codigo.matchAll(new RegExp(`${nombre}(?: = |, )\\{([\\s\\S]*?)\\n\\}`, 'g'))) {
    for (const par of bloque[1].matchAll(/:\s*'([^']+)'/g)) textos.add(par[1]);
  }
}

const faltan = [...textos].filter((s) => !(s in EN));
if (faltan.length) {
  console.log(`Faltan ${faltan.length} traducciones:\n`);
  for (const s of faltan) console.log(`  ${JSON.stringify(s)}`);
  process.exit(1);
}
console.log(`Todo traducido (${textos.size} textos).`);

File: tools/i18n_missing.py
"""Textos del panel y del admin que se ven en español y no tienen inglés.

El panel y el admin se traducen «por lo que se ve» (assets/i18n.js): después
de pintar, cada texto se cambia por su entrada del diccionario si la hay. Este
script busca en el código los textos que se pintan (entre etiquetas, en
atributos que se traducen y en cadenas entre comillas) y lista los que no
están en el diccionario.

Uso:
  python tools/i18n_missing.py panel        # lista
  python tools/i18n_missing.py admin --json # para rellenar a mano
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

sys.stdout.reconfigure(encoding='utf-8')
ROOT = Path(__file__).resolve().parent.parent

AREAS = {
    # ideas.js ya trae cada idea en los dos idiomas.
    'panel': (['panel/panel.js', 'panel/index.html'], 'panel/i18n.js'),
    'admin': (['admin/admin.js', 'admin/index.html'], 'admin/i18n.js'),
    'cuenta': (['app/app.js', 'app/cuenta.js'], 'app/i18n.js'),
}

# Palabras que delatan una frase en español.
ES = re.compile(r"[áéíóúñ¿¡]|\b(el|la|los|las|de|del|que|con|para|por|una?|sin|más|esta?|este|hay|tu|tus|se|ya|en|al|y|o|lo|es|son)\b", re.I)
# …y en inglés (los textos ya en inglés no se cuentan).
EN = re.compile(r"\b(the|and|to|of|your|you|is|are|with|for|in|on|this|it|at|an?|or|be|no|not|can|will|from|by)\b", re.I)
BILINGUE = re.compile(r"\b(en|EN)\s*\?|I18N\.lang === 'en'|lang === 'en'|\bbi\(|^\s*[?:] [`'\"\[(]")
# Nombres propios y palabras que son iguales en inglés.
MARCAS = {'Klendar', 'Supabase', 'Google', 'Apple', 'Stripe', 'Mapbox', 'Resend', 'Cloudflare', 'Madrid', 'Barcelona',
          'Pro', 'Error', 'Email', 'Web', 'Premium', 'Instagram', 'Facebook', 'Twitter', 'Android', 'Total', 'Normal',
          'Material', 'Local', 'General', 'Final', 'Hola', 'Enter', 'Escape', 'Tab'}
# Lo que no se enseña o ya va en los dos idiomas.
IGNORAR = {'sin configuración', 'Idioma / Language', 'Español', 'English', 'Invalid login credentials',
           'User already registered', 'Payload too large', 'Klendar · Panel del negocio', 'Klendar · Administración'}
CODIGO = re.compile(r"^[\w.-]+$|[{}=;()\[\]]|=>|\.(js|css|png|svg)\b|^#|^/|https?:|^\$|^[a-z_]+:|--|\\n")


def candidatos(texto: str):
    abiertos = 0  # paréntesis de un `bi(` que sigue en las líneas de abajo
    for linea in texto.splitlines():
        # Las líneas que ya eligen idioma a mano (`en ? '…' : '…'`, `bi(…)`).
        if abiertos > 0 or 'bi(' in linea:
            abiertos += linea.count('(') - linea.count(')')
            continue
        if BILINGUE.search(linea):
            continue
        # 1. Texto entre etiquetas (también dentro de plantillas `...`).
        for m in re.finditer(r">([^<>`]+?)<", linea):
            yield m.group(1)
        # 2. Trozos de plantilla pegados a una interpolación.
        for m in re.finditer(r">([^<>`$]+?)\$\{", linea):
            yield m.group(1)
        for m in re.finditer(r"\}([^<>`${]+?)<", linea):
            yield m.group(1)
        # 3. Atributos que se traducen.
        for m in re.finditer(r"(?:placeholder|title|aria-label|alt)=\"([^\"$]+)\"", linea):
            yield m.group(1)
        # 4. Cadenas entre comillas simples o dobles.
        for m in re.finditer(r"'((?:[^'\\]|\\.){2,})'", linea):
            yield m.group(1).replace("\\'", "'")
        for m in re.finditer(r"\"((?:[^\"\\]|\\.){2,})\"", linea):
            yield m.group(1)


def limpia(s: str) -> str | None:
    s = re.sub(r"\s+", " ", s).strip()
    s = s.strip('·—–:|,').strip()
    if len(s) < 2 or not re.search(r"[A-Za-zÁÉÍÓÚáéíóúñÑ]{2}", s):
        return None
    # Una sola palabra: botones y etiquetas («Guardar», «Borrar», «dentro»).
    # Con mayúscula, o en minúscula con tilde; el resto suele ser código.
    if re.fullmatch(r"[A-ZÁÉÍÓÚ][a-záéíóúñü]+|[a-záéíóúñü]*[áéíóúñ][a-záéíóúñü]*", s):
        return None if s in MARCAS else s
    if CODIGO.search(s) or '<' in s:
        return None
    es, en = len(ES.findall(s)), len(EN.findall(s))
    if en > es:
        return None
    if not es and not re.match(r"^[A-ZÁÉÍÓÚ¿¡][a-záéíóúñ]+(\s|$)", s):
        return None
    return s


def dic(path: Path) -> set[str]:
    t = path.read_text(encoding='utf-8')
    return {m[1] for m in re.findall(r"^\s*(['\"])(.+?)\1\s*:", t, re.M)}


def main():
    area = sys.argv[1] if len(sys.argv) > 1 else 'panel'
    fuentes, dpath = AREAS[area]
    ya = dic(ROOT / dpath)
    faltan: dict[str, str] = {}
    for f in fuentes:
        txt = (ROOT / f).read_text(encoding='utf-8')
        for c in candidatos(txt):
            s = limpia(c)
            if s and s not in ya and s not in faltan and s not in IGNORAR:
                faltan[s] = f
    if '--json' in sys.argv:
        print(json.dumps({k: '' for k in faltan}, ensure_ascii=False, indent=2))
    else:
        for k, f in faltan.items():
            print(f'{f}: {k}')
        print(f'\n{len(faltan)} texto(s) sin inglés en {area}.')


if __name__ == '__main__':
    main()
